pkg/kube/client.go

```diff
@@ -180,6 +185,9 @@ func (c *kubernetesClient) Init() error {
 		return err
 	}
 
+	c.cachedDiscoveryClient = memory.NewMemCacheClient(c.Discovery())
+	go wait.Forever(func() { c.cachedDiscoveryClient.Invalidate() }, 2*time.Minute)
```

Comment:
Great enhancement! I have some considerations:

1. Periodic invalidation is not always necessary.

- APIResourceList is called only on startup. So we can stop a periodic invalidation after all Monitors are started, e.g. if there are only conversion/validating hooks. 
- A constant period is not good, we've faced consequences already (see #226).

May be [JitterUntil](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#JitterUntil) or its buddy JitterUntilWithContext can give us more control over this cache mechanism.

2. Can we face a reincarnation of a problem with "non-working api controllers"? There is a NOTE in [docs](https://pkg.go.dev/k8s.io/client-go/discovery/cached/memory):
> NOTE: The client will NOT resort to live lookups on cache misses.

I can predict this scenario: "Invalidate" can't get GVRs → cache is empty → APIResourceList can't get GVR → shell-operator can't start informer for Pods.

I think we should elaborate more on this enhancement...

Reply:
Thanks for the list. I've referenced it (along with a link to this naïve implementation) in the [issue](https://github.com/flant/shell-operator/issues/254).

I've introduced this to possibly avoid useless API discovery when modifying Kubernetes objects. Clearly, the theoretical speed-up is not worth the hassle for now.

Let's scrap this feature for now since it affects the shell-operator in many ways I have not predicted.